packages/http/test/file.test.ts

```diff
@@ -0,0 +1,547 @@
+
+/*
+
+CASES TO TEST
+
+// `"single"` body: uploads file as a JSON structure { filename: <string>,
+// contentType: <string>, contents: <base64> }
+op uploadFileAsJson(@header contentType: "application/json", @body file: File)
+
+// `"file"` body: Uploads file as a binary data stream. Content-Type has contentType and
+// filename cannot be specified (Content-Disposition is a response-only header)
+op uploadFile(@body file: File);
+op uploadFile(...File);
+
+// `"file"` body: Downloads file as binary data stream. Content-Type has contentType and
+// filename comes from Content-Disposition
+op downloadFile(): { @body file: File };
+op downloadFile(): File;
+
+
+// Declares a custom textual file that can either be "application/json" or "application/yaml"
+model SpecFile extends File<Contents = bytes> {
+  contentType: "application/json" | "application/yaml";
+  // We can put HTTP metadata inside the file model as long as we use it with `@bodyRoot` and not
+  // `@body`      
+  @header("x-filename") filename: string;
+}
+
+// ??? with @header on contentType
+op uploadWeird(@bodyRoot file: SpecFile): void;
+
+// `"file"` body: Uploads spec as a file. Content-Type has contentType, x-filename has filename.
+// isText: true
+op uploadSpec(@bodyRoot spec: SpecFile): void;
+
+// `"single"` body: Uploads spec as a JSON object { contentType: <json or yaml>,
+// filename: <string>, contents: <string> }
+op uploadSpec(@header contentType: "application/json", @body spec: SpecFile): void;
+
+op uploadFileMultipart(@multipartBody fields: { file: HttpPart<SpecFile> }): void;
+
+op uploadFileNested(@body data: { file: File }): void;
+
+op uploadFileInPieces(
+  @header contentType: "application/json" | "application/yaml",
+  // filename?
+  @body contents: bytes,
+}: void;
+*/
+
+describe("http: file", () => {
+  it("does not allow instances that improperly provide contents argument", async () => {
+    const { diagnostics } = await compileOperationsFull(`
+      model BadFile1 extends Http.File<string | bytes> {}
+
+      model BadFile2 extends Http.File<"asdf"> {}
+    `);
+
+    expectDiagnostics(diagnostics, [
+      {
+        code: "@typespec/http/http-file-contents-not-scalar",
+        severity: "error",
+        message:
+          "The 'contents' property of the file model must be a scalar type that extends 'string' or 'bytes'. Found 'string | bytes'.",
+      },
+      {
+        code: "@typespec/http/http-file-contents-not-scalar",
+        severity: "error",
+        message:
+          "The 'contents' property of the file model must be a scalar type that extends 'string' or 'bytes'. Found '<value>'.",
+      },
+    ]);
+  });
+
+  it("exact payload upload and download", async () => {
+    const [
+      {
+        parameters: { body: requestBody },
+        responses: [
+          {
+            responses: [{ body: responseBody }],
+          },
+        ],
+      },
+    ] = await getOperations(`
+      op example(...Http.File): Http.File;
```

Comment:
what if the `File` is intersect with other things like `...CommonHeader & Http.File`?

Reply:
Good question. The current implementation will _not_ regard this as a File, and it's difficult for the HTTP layer to handle this because intersections are transparent in TypeSpec i.e. there is no Type with `"kind": "Intersection"`, we just see a model with all the properties of both models.

I think we can maybe detect if the File is in intersection with models that only contain HTTP metadata like `@statusCode`, but it will be pretty hard to do this correctly due to how intersections work.

Reply:
then i doubt current concept for the file will fail for some azure core template. in azure core, we always use intersection to provide trait features. see [example](https://github.com/Azure/typespec-azure/blob/b5393c58703ea0dd83fd3af62d8851c42ec26bd6/packages/typespec-azure-core/lib/operations.tsp#L33). for such cases, file will never be detected?

Reply:
another case is multiple spread case, like `op test(...File, ...CommonHeader): void`.

Reply:
I don't think current "composite" Azure templates are suitable for using `File` anyway. We would probably want to provide designated templates for operations that upload/download files, and probably also a designated model that extends File that Azure services use.

Yes, these spread/intersection cases are tough to implement. I did attempt to make this work yesterday but the logic is very complicated.

Reply:
Through @chrisradek's efforts, we've found a way to make cases like `OkResponse & File` as well as `...CommonHeader, ...File` work correctly with some not-so-complicated logic. The logic is: "if the model has all of the properties of `Http.File`, and after removing all applicable metadata it has _only_ properties of `Http.File`, then it's a file". This allows cases like you mentioned to still be recognized as files. I'm working on updating the docs for these cases now.

Reply:
Resolving as we now account for this and test these cases.